src/main/host/timer.rs

```diff
@@ -0,0 +1,237 @@
+        self.internal.borrow().expire_interval
+    }
+
+    pub fn disarm(&mut self) {
+        self.magic.debug_check();
+        let mut internal = self.internal.borrow_mut();
+        internal.next_expire_time = None;
+        internal.expire_interval = SimulationTime::ZERO;
+        internal.expiration_count = 0;
+        internal.min_valid_expire_id = internal.next_expire_id;
+    }
+
+    fn timer_expire(
+        internal_weak: &Weak<AtomicRefCell<TimerInternal>>,
+        host: &mut Host,
+        expire_id: u64,
+    ) {
+        let internal = if let Some(internal) = Weak::upgrade(internal_weak) {
+            internal
+        } else {
+            trace!("Expired Timer no longer exists.");
+            return;
+        };
+        let mut internal_brw = internal.borrow_mut();
+        trace!(
+            "timer expire check; expireID={} minValidExpireID={}",
+            expire_id,
+            internal_brw.min_valid_expire_id
+        );
+        if expire_id < internal_brw.min_valid_expire_id {
+            // Cancelled.
+            return;
+        }
+
+        let next_expire_time = internal_brw.next_expire_time.unwrap();
+        if next_expire_time > Worker::current_time().unwrap() {
+            // Hasn't expired yet. Check again later.
+            Self::schedule_new_expire_event(&mut *internal_brw, internal_weak.clone(), host);
+            return;
+        }
+
+        internal_brw.expiration_count += 1;
+        if internal_brw.expire_interval > SimulationTime::ZERO {
+            internal_brw.next_expire_time =
+                Some(internal_brw.next_expire_time.unwrap() + internal_brw.expire_interval);
+            Self::schedule_new_expire_event(&mut *internal_brw, internal_weak.clone(), host);
+        }
+
+        // Re-borrow as an immutable reference while executing the callback.
+        drop(internal_brw);
+        let internal_brw = internal.borrow();
+        (internal_brw.on_expire)(host);
+    }
+
+    fn schedule_new_expire_event(
+        internal_ref: &mut TimerInternal,
+        internal_ptr: Weak<AtomicRefCell<TimerInternal>>,
+        host: &mut Host,
+    ) {
+        let now = Worker::current_time().unwrap();
+        let delay = std::cmp::min(
+            internal_ref.next_expire_time.unwrap() - now,
+            SimulationTime::SECOND,
+        );
+        let expire_id = internal_ref.next_expire_id;
+        internal_ref.next_expire_id += 1;
+        let task = TaskRef::new(move |host| Self::timer_expire(&internal_ptr, host, expire_id));
+        Worker::schedule_task_with_delay(task, host, delay);
+    }
+
+    pub fn arm(
+        &mut self,
+        host: &mut Host,
+        expire_time: EmulatedTime,
+        expire_interval: SimulationTime,
+    ) {
+        self.magic.debug_check();
+
+        let mut internal = self.internal.borrow_mut();
+        debug_assert!(expire_time >= Worker::current_time().unwrap());
+        internal.next_expire_time = Some(expire_time);
```

Comment:
What happens in release builds if the expire time is less? Does the scheduler handle this reasonably?

Reply:
Yes, the scheduler allows it and things generally work.